[18.0][MIG] web_timeline

JasminSForgeFlow opened the pull request:
Standard Migration

Depends on:

- https://github.com/OCA/web/pull/3145 For tests

@ForgeFlow

lmignon committed:
[IMP] web_timeline: Extracted from incubator

Commit:
[IMP] web_timeline: Several improvements:

* Avoid to display all items of group_by model
* Remove unnecessary readgroup
* Remove dependency on project. Modify module structure. Imporve readme file.
* Add setup.py

hbrunn committed:
[IMP] web_timeline: add few features

pedrobaeza committed:
[MIG] web_timeline: Migration to 9.0

ThomasBinsfeld committed:
[FIX] web_timeline: fix commas in manifest file

JBF91 committed:
[10.0] [FIX] [web_timeline] Correct write when grouped by (#678)

* [FIX] Correct write when grouped by

* Remove blank lines

* Minor changes

Creamaster committed:
[10.0][IMP] web_timeline improvements:

* Update the vis.js framework to latest stable version (v4.20.1) to add new functionality and save almost
400ko of source code in odoo assets (the vis dev team has isolate the code for timeline widget from the
 rest of their widgets)

* Add the possibility to display events with no duration (instantaneous) as the same time has regular events

* Add the possibility to choose between two scroll mode :
  -  zoom in time resolution (default)
  or
  -  vertical scroll (web browser default behaviour)

Signed-off-by: adrien.didenot <adrien.didenot@horanet.com>

Creamaster committed:
[Fix OCA/project#317] Add the property date_delay to TimeLine views (same as base Calendar view)

In Odoo calendar view, the attribute date_delay is an alternative to date_stop, to provides the duration of the event instead of its end date.
This commit integrate this attribute to the Timeline view, but contrary to the base Calendar view the Timeline view could use both arguments (date_stop and date_delay) simultaneously.

Signed-off-by: adrien.didenot <adrien.didenot@horanet.com>

Creamaster committed:
[IMP] Add an option to initialize the timeline window on display.

The default window display all the events (aka 'fit'), in case of events spread over the year, the events are invisibles (too small to be readable)

Signed-off-by: adrien.didenot <adrien.didenot@horanet.com>

Creamaster committed:
[IMP] change the option to initialize the timeline window on display.

Change the attribute 'default_window' to 'mode' like the Odoo calendar view (to be iso functional)

npiganeau committed:
Fixed web_timeline.css

CSS classes have changed with the updated version of vis.js

Maartincm committed:
[MIG] web_timeline module from v10 to v11

* Update Version in Manifest
* Remove enconding in .py files
* Rewrite the view definition according to version 11

Maartincm committed:
[FIX] Missing GroupClick Functionality

tarteo committed:
[FIX] web_timeline: Make colors attribute optional (#969)

tarteo committed:
[IMP] web_timeline: Templates implementation like kanban

[IMP] web_timeline: Templates implementation like kanban

[ADD] Check if template exists

Update README.rst

tarteo committed:
[ADD] field_utils to web_timeline

[UPD] Version number

[UPD] README

tarteo committed:
[ADD] web_timeline: New dependency_arrow attribute

Update README.rst

[FIX] Remove console.log

[ADD] Make timeline.fit optional

[FIX] Use stringified points

[IMP] Reversed the arrow head and fixed lint issues

[IMP] Use options parameter for line color and width

[FIX] Version number

[IMP] Minor improvements

tarteo committed:
[ADD] Enable multiselect

[IMP] Prevent flickering

[IMP] Fixed headers

[ADD] Create new record by date range selection

Remove roadmap item from readme

[FIX] minimal height

tarteo committed:
[MIG] web_timeline: Migration to 12.0

tarteo committed:
[IMP] Readme: Available attribute list

Commit:
[IMP] Add stack support

Commit:
[FIX] web_timeline: fix readme - To be squashed with previous one on migration (#1137)

cvinh committed:
[web_timeline][IMP] Add margin support in web_timeline (#1135)

tarteo committed:
[FIX] web_timeline: Reloading view

[FIX] web_timeline: Reloading view

[FIX] web_timeline: Reloading view

denislour committed:
[12.0][FIX]web_timeline: fix group orders

denislour committed:
[12.0][web_timeline] Fix bug from group_order: locale instead locate.

hpatelserpentcs committed:
[FIX][12.0] web_timeline: Scroll Issue.

thoongnv committed:
[IMP] web_timeline: black, isort

thoongnv committed:
[MIG] web_timeline: Migration to 13.0

Tardo committed:
[MIG] web_timeline: Finish migration to 13.0

Tardo committed:
[FIX] web_timeline: Change vis-timeline library import

Tardo committed:
[IMP] web_timeline: CSS vis items

Tardo committed:
[FIX] web_timeline: Change draw_dependencies

Murtaza-OSI committed:
[FIX] Travis

Murtaza-OSI committed:
[14.0][MIG] Migrated web_timeline.

cesar-tecnativa committed:
[MIG] web_timeline: Migration to 15.0

hbrunn committed:
[FIX] web_timeline: Don't crash when restoring timeline from breadcrumb

CarlosRoca13 committed:
[FIX] web_timeline: Allow groupby m2m fields

Commit:
[IMP] web_timeline: allow UNASSIGNED grouping when using M2M field

alexis-via committed:
[FIX] web_timeline: order group_by field

Fix bug #2266: before this fix, the lines of the timeline view were always sorted by alphabetical order, ignoring the native order of the model on which the group_by field points to.

Fix duplicate entries in the 'fields' argument when the 'colors' parameters uses the same field in multiple conditions.

alexis-via committed:
Update vis-timeline-graph2d lib: 7.3.4 -> 7.7.0

Commit:
[UPD] Update web_timeline.pot

jguenat committed:
[FIX] web_timeline: utc and create_completed
fix wrong time used when creating new record

on record creation timeline view was reloaded and all items disappeared
fix by removing unnecessary reload

jguenat committed:
[FIX] web_timeline: updateProm.finally is not a function
Following odoo#102621 an UncaughtPromiseError > TypeError occured each time an update was triggered in timeline view
Fixed by returning super

bvanjeelharia committed:
[MIG] web_timeline: Migration to 16.0

Syntax changes

Syntax changes

tate11 committed:
Added translation using Weblate (Turkish)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (13 of 13 strings)

Translation: web-16.0/web-16.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_timeline/hr/

Yadier-Tecnativa committed:
[FIX] web_timeline: css and styles not charged when rendering timeline

On library version update a new feature was added at Timeline creation level to
allow modifying the xss. By keeping it active the css classes are not showed

As the view is on backend, has no sense to keep it active. So we deactivate the
feature to allow show the timeline correctly.

TT43238

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (13 of 13 strings)

Translation: web-16.0/web-16.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_timeline/es/

norlinhenrik committed:
[FIX] web_timeline: Add record - FormViewDialog - resId: null -> false (cannot be null)

norlinhenrik committed:
[FIX] web_timeline: date_delay = date_stop - date_start

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: web-16.0/web-16.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_timeline/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (13 of 13 strings)

Translation: web-16.0/web-16.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_timeline/pt_BR/

mymage committed:
Added translation using Weblate (Italian)

houzefa-abba committed:
[FIX] web_timeline: Wrap fields_get arg into a list

By contract the first argument of the `fields_get` method is supposed to
be a list. Before this fix, `web_timeline` would call `fields_get` with
a string instead of a list.

Fortunately in case only 1 field is being grouped, this worked as Odoo
does an `x in y` comparison in its `fields_get` implementation, which
does pass in a simple `"project_id" in "project_id"` case.

But that call remains invalid and can break when `fields_get` has been
tweaked by other modules.

houzefa-abba committed:
Translated using Weblate (French)

Currently translated at 100.0% (13 of 13 strings)

Translation: web-16.0/web-16.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-16-0/web-16-0-web_timeline/fr/

houzefa-abba committed:
[IMP] web_timeline: Overflow text, add hover colors

When an item label does not fit in its date-range box, overflow
according to
https://visjs.github.io/vis-timeline/examples/timeline/items/rangeOverflowItem.html

Previous CSS code was already trying to do that, but was selecting
`.vis-item.vis-item-content` instead of `.vis-item .vis-item-content`.

Displaying overflow text brings up layout issues solved by removing the
forced-100% width instruction.

This change also adds highlight when hovering a box, which is useful on
text that has overflown (as it has no borders).

houzefa-abba committed:
[IMP] web_timeline: Handle inverted date ranges

Inverted dates make no functional sense, but when they do happen, this
change allows displaying them as single points in the timeline (same as
when begin=end).

We currently do have such demo data when installing project_timeline.

houzefa-abba committed:
[IMP] web_timeline: Add demo view on cron tasks

This way we can showcase this module without project_timeline.

houzefa-abba committed:
[FIX] web_timeline: Redraw issues with initial mode

This commit fixes redraw issues when setting a `mode` attribute in the
`timeline` view tag.

This mode specifies a default scale one would want to set; same as when
clicking on Day/Week/Month buttons at the top of the view.

Initial rendering had issues here because data was loaded too soon,
before the timeline component was rendered/ready. The fix is to load
data into the component only after initial redraw event, called
`changed` (see <https://visjs.github.io/vis-timeline/docs/timeline/#Events>).

There was old code attempting to call `on_scale_xxx_clicked` methods at
load time to simulate clicks on these Day/Week/Month buttons, but these
methods have been renamed so this code is no longer working.

This commit also removes the `current_window` instance variable, not
needed and actually confusing as the timeline component already
maintains its own start/end information (which we can query with
`timeline.getWindow()`).

houzefa-abba committed:
[IMP] web_timeline: Align to top instead of bottom

This avoids somewhat ugly very large "Unassigned" first row.

houzefa-abba committed:
[IMP] web_timeline: Follow create/edit/delete attrs

In addition to security rights (was already implemented), now follow
`create="0"` / `edit="0"` / `delete="0"` attributes one can set onto the
`timeline` tag, same as in other Odoo views.

carlos-lopez-tecnativa committed:
[IMP] web_timeline: pre-commit auto fixes

carlos-lopez-tecnativa committed:
[MOVE] web_timeline: new folder structure according Odoo guideslines

carlos-lopez-tecnativa committed:
[MIG] web_timeline: Migration to 17.0
- Convert Moment.js to Luxon.
- Replace Underscore.js with native JavaScript code.
- Migrate legacy views to the new system and add an architecture parser to separate logic.
- added basic test

Commit:
[UPD] Update web_timeline.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: web-17.0/web-17.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-17-0/web-17-0-web_timeline/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (16 of 16 strings)

Translation: web-17.0/web-17.0-web_timeline
Translate-URL: https://translation.odoo-community.org/projects/web-17-0/web-17-0-web_timeline/it/

carlos-lopez-tecnativa committed:
[IMP] web_timeline: Update vis-timeline-graph2d lib: 7.7.0 -> 7.7.3

OCA-git-bot committed:
[BOT] post-merge updates

JasminSForgeFlow committed:
[IMP] web_timeline: black, isort, prettier

JasminSForgeFlow committed:
[MIG] web_timeline: Migration to 18.0

carlos-lopez-tecnativa commented on web_timeline/static/src/views/timeline/timeline_controller.esm.js:
```diff
@@ -0,0 +1,273 @@
+
+const {DateTime} = luxon;
+
+// Import time from "web.time";
```
Please remove this line

JasminSForgeFlow replied:
Removed, Thanks

carlos-lopez-tecnativa commented on web_timeline/tests/test_web_timeline.py:
```diff
@@ -0,0 +1,25 @@
+class TestWebTimeline(HttpCase):
+    def test_timeline_arch(self):
+        self.browser_js(
+            "/web/tests/legacy?mod=web&filter=TimelineView - ArchParser",
```
Is it complicated to convert the tests to the new style?

JasminSForgeFlow replied:
not sure about it need to take a look

carlos-lopez-tecnativa commented on web_timeline/tests/test_web_timeline.py:
```diff
@@ -0,0 +1,25 @@
+
+    def test_timeline_view(self):
+        self.browser_js(
+            "/web/tests/legacy?mod=web&filter=TimelineView - View",
```
the same here

carlos-lopez-tecnativa reviewed (changes requested):
Tested locally with https://github.com/OCA/project/pull/1478.
Please check the CI or reference the PR to solve the issue in the test.

TT56064

pedrobaeza commented:
/ocabot migration web_timeline

pedrobaeza commented:
@JasminSForgeFlow any news?

pedrobaeza commented:
CI is still red.

JasminSForgeFlow commented:
> CI is still red.

It's because of this warning

`2025-05-09 06:39:17,536 287 WARNING odoo odoo.addons.web_timeline.tests.test_web_timeline.TestWebTimeline.test_timeline_view.browser: You disabled XSS protection for vis-Timeline. I sure hope you know what you're doing! `

which is deactivated https://github.com/OCA/web/commit/e3e9b5d4ee3a0b583ae9f1ae21ab863159e935fd

pedrobaeza commented:
@carlos-lopez-tecnativa do you know if this disabling is still needed? We should find any way of avoiding the red CI, silencing the warning or removing it at all.

carlos-lopez-tecnativa committed:
[FIX] web_timeline: Enable XSS filtering with whitelist for safer HTML rendering

Before this commit, XSS filtering was disabled, which could introduce potential security risks.
This commit adopts a safer approach, similar to PR https://github.com/OCA/web/pull/2525.
After this commit, XSS filtering is re-enabled, and a whitelist of allowed HTML elements is provided.

carlos-lopez-tecnativa commented:
@JasminSForgeFlow please include https://github.com/OCA/web/pull/3169 to test that the XSS issue is resolved.

JasminSForgeFlow commented:
> @JasminSForgeFlow please include #3169 to test that the XSS issue is resolved.

Done, Thanks

carlos-lopez-tecnativa committed:
[FIX] web_timeline: Add <b> element to XSS whitelist for safer HTML rendering

carlos-lopez-tecnativa commented:
Please include https://github.com/OCA/web/pull/3171

JasminSForgeFlow commented:
> Please include #3171

Added

carlos-lopez-tecnativa reviewed (approved):
LGTM @CarlosRoca13 @pedrobaeza please review this

CarlosRoca13 reviewed (approved):
/ocabot merge nobump
